source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +329,160 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
             return result;
         }
 
-
-        void AppendRelatedDocumentStatementsForInsert(
-            StringBuilder sb,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateInsertCommandsWithRelatedDocuments(
+            StringBuilder insertStatementBuilder,
             DocumentMap mapping,
-            IReadOnlyList<object> documents)
+            IReadOnlyList<object> documents,
+            InsertOptions options)
         {
+            var commands = new List<PreparedCommand>();
+            var parameters = new CommandParameterValues();
+            // we need to create these document parameters for each command - parameters are disposed at the end of execution in 'CommandExecutor'
+            Func<CommandParameterValues> documentParametersCreator = () => GetDocumentParameters(m => keyAllocator(m), options.CustomAssignedId, documents, mapping, DataModification.Insert);
+            parameters.AddRange(documentParametersCreator());
+
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, documents);
+            if (!relatedDocumentData.Any(x => x.Related.Any()))
+            {
+                commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+                return commands.ToArray();
+            }
 
             foreach (var data in relatedDocumentData.Where(g => g.Related.Length > 0))
             {
                 var relatedVariablePrefix = $"{data.TableName.ToLower()}_";
 
-                sb.AppendLine($"INSERT INTO [{data.SchemaName}].[{data.TableName}] ([{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}]) VALUES");
-                var related = data.Related;
-
-                for (var x = 0; x < related.Length; x++)
+                var remaining = data.Related.Length;
+                while (remaining > 0)
                 {
-                    var parentIdVariable = related[x].parentIdVariable;
-                    var relatedDocumentId = related[x].relatedDocumentId;
-                    var relatedTableName = related[x].relatedTableName;
-
-                    var relatedVariableName = relatedVariablePrefix + x;
-                    parameters.Add(relatedVariableName, relatedDocumentId);
-                    if (x > 0)
-                        sb.Append(",");
-                    sb.AppendLine($"(@{parentIdVariable}, '{mapping.TableName}', @{relatedVariableName}, '{relatedTableName}')");
+                    var reachedParameterLimit = false;
+
+                    while (!reachedParameterLimit && remaining > 0)
```

Comment:
It looks like we are adding 1 new parameter per related document? So could we split and process each batch separately rather than using a while loop?

Reply:
I don't really follow what you're asking here. What's the improvement over the batching structure that exists here?

On a side note, this function doesn't need the second while loop defined anymore as I no longer use the `reachedParameterLimit` variable so that drops a few lines of code.

Reply:
Sorry should've been more specific: I'm thinking that readability/understandability would be improved by not using a while loop.

Could you push up those changes now anyway?

Reply:
Pushed

Reply:
Closing this conversation so it's not confusing looking at old code with whatever suggestion is made.